pkg/kubelet/volumemanager/reconciler/reconciler.go

```diff
@@ -69,180 +80,168 @@ func NewReconciler(
 	}
 }
 
 type reconciler struct {
 	kubeClient                    internalclientset.Interface
 	controllerAttachDetachEnabled bool
 	loopSleepDuration             time.Duration
+	reconstructDuration           time.Duration
 	waitForAttachTimeout          time.Duration
 	hostName                      string
 	desiredStateOfWorld           cache.DesiredStateOfWorld
 	actualStateOfWorld            cache.ActualStateOfWorld
 	operationExecutor             operationexecutor.OperationExecutor
 	mounter                       mount.Interface
+	volumePluginMgr               *volume.VolumePluginMgr
+	kubeletPodsDir                string
+	timeOfLastReconstruct         time.Time
 }
 
-func (rc *reconciler) Run(stopCh <-chan struct{}) {
-	wait.Until(rc.reconciliationLoopFunc(), rc.loopSleepDuration, stopCh)
+func (rc *reconciler) Run(sourcesReady config.SourcesReady, stopCh <-chan struct{}) {
+	wait.Until(rc.reconciliationLoopFunc(sourcesReady), rc.loopSleepDuration, stopCh)
 }
 
-func (rc *reconciler) reconciliationLoopFunc() func() {
+func (rc *reconciler) reconciliationLoopFunc(sourcesReady config.SourcesReady) func() {
 	return func() {
-		// Unmounts are triggered before mounts so that a volume that was
-		// referenced by a pod that was deleted and is now referenced by another
-		// pod is unmounted from the first pod before being mounted to the new
-		// pod.
+		rc.reconciliation()
 
-		// Ensure volumes that should be unmounted are unmounted.
-		for _, mountedVolume := range rc.actualStateOfWorld.GetMountedVolumes() {
-			if !rc.desiredStateOfWorld.PodExistsInVolume(mountedVolume.PodName, mountedVolume.VolumeName) {
-				// Volume is mounted, unmount it
-				glog.V(12).Infof("Attempting to start UnmountVolume for volume %q (spec.Name: %q) from pod %q (UID: %q).",
+		// Add all sources ready check so that reconciler's reconstruct process will start after
+		// desired state of world is populated with pod volume information from different sources. Otherwise,
+		// reconciler's reconstruct process may add incomplete volume information and cause confusion.
+		// In addition, if some sources are not ready, the reconstruct process may clean up pods' volumes
+		// that are still in use because desired states could not get a complete list of pods.
+		if sourcesReady.AllReady() && time.Since(rc.timeOfLastReconstruct) > rc.reconstructDuration {
+			glog.V(5).Infof("Sources are all ready, starting reconstruct state function")
+			rc.reconstruct()
+		}
+	}
+}
+
+func (rc *reconciler) reconciliation() {
+	// Unmounts are triggered before mounts so that a volume that was
+	// referenced by a pod that was deleted and is now referenced by another
+	// pod is unmounted from the first pod before being mounted to the new
+	// pod.
+
+	// Ensure volumes that should be unmounted are unmounted.
+	for _, mountedVolume := range rc.actualStateOfWorld.GetMountedVolumes() {
+		if !rc.desiredStateOfWorld.PodExistsInVolume(mountedVolume.PodName, mountedVolume.VolumeName) {
```

Comment:
@jingxu97 @saad-ali I think if we restructure the code in lines 153 to 177 then we might get more readable and simpler flow. Let me show you tomorrow in person. It will only take a second.